### DISCUSSION TITLE
Having multiple nuget feeds - only private feed is queried

### DISCUSSION THREAD

**Discussion by kipusoep** [renovatebot/renovate#14782] [Request Help]
When you have the following config (partial):
```json
"nuget": {
  "registryUrls": [
    "https://api.nuget.org/v3/index.json",
    "https://pkgs.dev.azure.com/<redacted>/<redacted>/_packaging/<redacted>/nuget/v3/index.json"
  ]
}
```
Only the private feed seems to be used to search for updates. The problem with that, is that public packages are listed later in the private feed than in the public feed. If you combine that with the `stabilityDays` property, it's possible that packages that have been published earlier than the current date - `stabilityDays` days, won't get elected for updating.

How can I make sure that the public feed is used for public packages?
Complete config:

```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "enabledManagers": [
        "nuget"
    ],
    "rebaseWhen": "behind-base-branch",
    "labels": [
        "Renovate",
        "BackEnd"
    ],
    "major": {
        "enabled": false
    },
    "packageRules": [
        {
            "matchPackagePrefixes": [
                "Redacted."
            ],
            "groupName": "all internal dependencies",
            "groupSlug": "all-internal"
        },
        {
            "excludePackagePrefixes": [
                "Redacted."
            ],
            "matchUpdateTypes": [
                "minor",
                "patch",
                "pin",
                "digest"
            ],
            "stabilityDays": 7,
            "prCreation": "not-pending",
            "groupName": "all external non-major dependencies",
            "groupSlug": "all-external-non-major"
        }
    ],
    "hostRules": [
        {
            "platform": "nuget",
            "endpoint": "https://pkgs.dev.azure.com/Redacted/Redacted/_packaging/Redacted/nuget/v3/index.json",
            "username": "",
            "password": "***********"
        },
        {
            "hostType": "github",
            "matchHost": "github.com",
            "token": "***********"
        }
    ],
    "azureWorkItemId": 12345,
    "nuget": {
        "registryUrls": [
            "https://api.nuget.org/v3/index.json",
            "https://pkgs.dev.azure.com/Redacted/Redacted/_packaging/Redacted/nuget/v3/index.json"
        ]
    },
    "repositories": [
        "Redacted/Redacted"
    ],
    "platform": "azure",
    "endpoint": "https://dev.azure.com/Redacted/",
    "token": "***********"
}
```

Output from command:
```
DEBUG: Update has not passed stability days (repository=Redacted/Redacted, branch=renovate/all-external-non-major)
       "depName": "Azure.Messaging.ServiceBus",
       "daysElapsed": 4,
       "stabilityDays": 7
DEBUG: Update has not passed stability days (repository=Redacted/Redacted, branch=renovate/all-external-non-major)
       "depName": "Azure.Storage.Files.DataLake",
       "daysElapsed": 4,
       "stabilityDays": 7
DEBUG: Update has not passed stability days (repository=Redacted/Redacted, branch=renovate/all-external-non-major)
       "depName": "Microsoft.VisualStudio.Azure.Containers.Tools.Targets",
       "daysElapsed": 3,
       "stabilityDays": 7

DEBUG: Skipping branch creation due to internal status checks not met (repository=Redacted/Redacted, branch=renovate/all-external-non-major)
```

The listed packages above have all been published more than 7 days ago, but not inside our private feed (which is shadow copying the public nuget feed by default, it's good to have it as a back-up for when NuGet.org is down).

Is there something I can do to work around this?

**Comment by rarkins** [comment]
Nuget's registry strategy is "merge": https://github.com/renovatebot/renovate/blob/fcaaaabbee3550d083b88c729c476dcafc1b779b/lib/modules/datasource/nuget/index.ts#L19

This means that it's meant to try all feeds and then merge the results. But you reported:

> Only the private feed seems to be used to search for updates. 

How sure are you about that?

**Reply by kipusoep** [reply]
Thanks for the reply!
I'm not sure to be honest, but it's using the 'wrong' result for my case.
Do you know how it's determining which result is used?
Or maybe some kind of workaround?

**Reply by rarkins** [reply]
Do you mean that the same version is found in *both*? And you want Renovate to prioritize one over the other?

**Reply by kipusoep** [reply]
Yeah the private feed is using the official nuget feed as upstream and saves a copy of the package. The publish date is just not the same, it's the date the package was cached.
I don't know if it's renovate's responsibility to prioritize, but a workaround would be nice 😅

**Reply by rarkins** [reply]
Here is the logic currently used for merging: https://github.com/renovatebot/renovate/blob/d1b13624c2939296d61ae9548d6617bf7a3e59f0/lib/modules/datasource/index.ts#L132-L184

Looking at that, if a release is found multiple times, the "first" one wins. This would mean the first registry URL's result is used, which would mean your private registry. If you want the public one to take precedence, can you try reversing the order of `registryUrls`?

**Reply by kipusoep** [reply]
Switching the positions of the feeds in the `RENOVATE_NUGET` env var doesn't have any effect unfortunately.

Before:
```
- name: RENOVATE_NUGET
  value: '{"registryUrls":["https://api.nuget.org/v3/index.json","$(vstsFeedSource)"]}'
```

After:
```
- name: RENOVATE_NUGET
  value: '{"registryUrls":["$(vstsFeedSource)","https://api.nuget.org/v3/index.json"]}'
```

**Reply by rarkins** [reply]
Unfortunately I don't think I can speculate any further unless you're able to reproduce the problem publicly somehow.

**Reply by kipusoep** [reply]
Yeah I was afraid that would be the next step. I don't know how to create a public reproduction sample without having to compromise our private feed. I guess we'll just have to 'live with the fact' that packages will be updated a few days later than we'd like to, it's not a big issue.

**Reply by rarkins** [reply]
Another way of looking at it: if you have a private feed, and it has "wrong" timestamps, then stabilityDays will be affected

**Reply by Andrioden** [reply]
My organization have the same problem, the minimumReleaseAge is counted from the internal nuget feed's date and not the public date. We use our internal feed as an [Azure upstream source](https://learn.microsoft.com/en-us/azure/devops/artifacts/concepts/upstream-sources?view=azure-devops#advantages) for security reasons. Since upstream source version are only updated on demand, it will always be a difference in time between the public feed date and internal feed date.

Changing the order of the feeds do not help
``` json
{
  nuget: {
    "registryUrls": [
      "https://pkgs.dev.azure.com/<redacted>/_packaging/<redacted>/nuget/v3/index.json",
      "https://api.nuget.org/v3/index.json"
    ]
  }
}
```

Is there any chance you give us a way to use the lowest date? Or specify which nuget feed should have priority?

**Example package link**
- https://www.nuget.org/packages/coverlet.collector
- Take note 6.0.4 was released 2025-02-19

**Example from logs tested with 39.180.2**
``` json
            {
             "deps": [
               {
                 "datasource": "nuget",
                 "depType": "nuget",
                 "depName": "coverlet.collector",
                 "currentValue": "6.0.3",
                 "registryUrls": [
                   "[https://pkgs.dev.azure.com/<redacted>/_packaging/<redacted>/nuget/v3/index.json"](https://pkgs.dev.azure.com/<redacted>/_packaging/<redacted>/nuget/v3/index.json%22)
                 ],
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "6.0.4",
                     "newValue": "6.0.4",
                     "releaseTimestamp": "2025-01-27T12:00:26.000Z",   // <--------------- Not 2025-02-19
                     "newVersionAgeInDays": 29,
                     "newMajor": 6,
                     "newMinor": 0,
                     "newPatch": 4,
                     "updateType": "patch",
                     "pendingChecks": true,
                     "libYears": 0.03553462709284627,
                     "branchName": "renovate/coverlet.collector-6.x"
                   }
                 ],
                 "packageName": "coverlet.collector",
                 "versioning": "nuget",
                 "warnings": [],
                 "sourceUrl": "https://github.com/coverlet-coverage/coverlet",
                 "registryUrl": "[https://pkgs.dev.azure.com/<redacted>/_packaging/<redacted>/nuget/v3/index.json",](https://pkgs.dev.azure.com/<redacted>/_packaging/<redacted>/nuget/v3/index.json%22,)
                 "currentVersion": "6.0.3",
                 "currentVersionTimestamp": "2025-01-14T12:43:26.000Z",
                 "currentVersionAgeInDays": 42,
                 "isSingleVersion": false
               }
```
